From e7d10d5472998f31ed76f4a3be901c86b1f9fe7e Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 11:59:57 -0400
Subject: [ethernet] mlx4: Use union for BlueFlame WQE

Message-id: <50ce75c724351857f63f995c1ae0a253cf8ebe73.1410425016.git.avadai@redhat.com>
Patchwork-id: 93623
O-Subject: [Patch RHEL7.1 V1 061/197] net/mlx4_en: Use union for BlueFlame WQE
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Amir Vadai <amirv@mellanox.com>

When BlueFlame is turned on, control segment of the TX WQE is changed,
and the second line of it is used for QPN.
Changed code to use a union in the mlx4_wqe_ctrl_seg instead of casting.
This makes the code clearer and solves the static checker warning:

drivers/net/ethernet/mellanox/mlx4/en_tx.c:839 mlx4_en_xmit()
 warn: potential memory corrupting cast 4 vs 2 bytes

CC: Dan Carpenter <dan.carpenter@oracle.com>
Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit ec5709403e6893acb4f7ca40514ebd29c3116836)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_tx.c b/drivers/net/ethernet/mellanox/mlx4/en_tx.c
index 77e8999..9cdeca4 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_tx.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_tx.c
@@ -875,7 +875,8 @@ netdev_tx_t mlx4_en_xmit(struct sk_buff *skb, struct net_device *dev)
  skb_tx_timestamp(skb);
 
  if (ring->bf_enabled && desc_size <= MAX_BF && !bounce && !vlan_tx_tag_present(skb)) {
-  *(__be32 *) (&tx_desc->ctrl.vlan_tag) |= cpu_to_be32(ring->doorbell_qpn);
+  tx_desc->ctrl.bf_qpn |= cpu_to_be32(ring->doorbell_qpn);
+
   op_own |= htonl((bf_index & 0xffff) << 8);
   /* Ensure new descirptor hits memory
   * before setting ownership of this descriptor to HW */
diff --git a/include/linux/mlx4/qp.h b/include/linux/mlx4/qp.h
index 59f8ba8..b66e761 100644
--- a/include/linux/mlx4/qp.h
+++ b/include/linux/mlx4/qp.h
@@ -270,9 +270,14 @@ enum {
 
 struct mlx4_wqe_ctrl_seg {
  __be32   owner_opcode;
- __be16   vlan_tag;
- u8   ins_vlan;
- u8   fence_size;
+ union {
+  struct {
+   __be16   vlan_tag;
+   u8   ins_vlan;
+   u8   fence_size;
+  };
+  __be32   bf_qpn;
+ };
  /*
   * High 24 bits are SRC remote buffer; low 8 bits are flags:
   * [7]   SO (strong ordering)
-- 
1.7.1